### ISSUE TITLE
[Feature Request] Add library API to schedule compute time steps

### ISSUE THREAD

**Issue by rbberger** [lammps/lammps#4360]
**Summary**

Add library API for calling `addstep` on computes.

**Detailed Description**

Some computes like `stress/atom` require the virial to be computed in the current time step.

When accessing such a compute via the library interface (`extract_compute`) at an arbitrary time step (nothing that matches up with either thermo or dump), there is currently no way to add such a step to the compute (via `addstep`).

Without the added step, the integration loop will not schedule the virial computation. Since we would like to run our Python analysis independent of both thermo and dump time steps, we would need library API methods to accomplish this.

@sjplimp what do you think?

**Further Information, Files, and Links**

**Comment by sjplimp** [comment]
I think it would be good to enable this via the lib interface.
Can you explain your use case in more detail (are you not doing integration steps at all)?
  E.g.  I think this would work fine if you were doing "run 0" repeatedly.
Also, what library calls are you proposing?

Steve

**Comment by rbberger** [comment]
@sjplimp this is a large simulation with a online local Python analysis happening at some interval that is independent of thermo and dumps disabled or very infrequent due to the large data volume.

**Comment by sjplimp** [comment]
I think what I don't understand is this:
The Python lib interface code wants to compute the stress/atom for a snapshot of atoms on a particular timestep.
(a) Where is it getting this snapshot from?
To do this it has to invoke the pair style on the snapshot (and build a neighbor list, etc).
(b) Assuming it isn't doing normal step-to-step time integration, why doesn't it do this via a "run 0"
     command, which would build the neigh list, invoke the pair style, and set the addstep flags to ensure
     the pair style also tallied virial info, so that compute stress/atom could be invoked after the "run 0"

Also I don't understand what you mean by an "online local Python analysis" - possibly you mean it is running
at the same time as the large simulation is running?  If that is the case, my (a) question is relevant.
And am I correct that the Python code still needs to invoke the pair style on the snapshot?

**Comment by rbberger** [comment]
The Python code runs as part of the regular integration loop at fixed intervals, at end_of_step (fix python/invoke). That's the "online" part. Pair style computation has happened and the neighbor list is accessible. What is missing is the necessary addstep for compute stress/atom so that the virial tally gets scheduled.

So the library API I'm looking for is being able to call addstep for a specific compute. Each Python invocation would then add the next scheduled step. fix invoke/python can't do this since it doesn't know which computes are needed.

I don't understand what you mean with snapshot. Each Python invocation has access to the entire local data of each MPI rank. If there are N MPI ranks, each of them is running its own local Python. So there are N Python instances in total with access to the corresponding MPI rank data. Instead of dumping data and analyzing it globally, each MPI rank analyzes its local subdomain while the simulation is running. That's the "local" part.

**Comment by sjplimp** [comment]
Hmmm - I posted this yesterday, but it didn't appear, not sure why.  Trying again.

Thanks for the additional info - I understand your workflow better now.
The way I would do this would not be thru Python and the lib interface.
But rather to add some code to the fix_python_invoke.cpp file.
Similar to what is in FixAveTime::invoke_scalar() or FixAdapt::change_settings().

In FixPythonInvoke::end_of_step() and post_force(), something like this:
  modify->clearstep_compute();
  call Python function
  modify->addstep_compute(update->ntimestep + nevery);

Also, you probably need to do something like FixAveTime::init() does to call
  modify->addstep_compute_all(nvalid);
so that the first call to FixPythonInvoke::init() can ensure the timestep for
the initial call to end_of_step() is flagged.

Basically this is treating fix python/invoke like the other 2 fixes as
one that can potentially invoke a compute (via Python code in this
case) that needs the virial (or energy) info from the pair style to be
tallied on particular timesteps.

What do you think of this approach?

**Comment by rbberger** [comment]
That will solve the python/invoke case. What about the case of having some Python driver that just does some arbitrary `run X` and then retrieving information of the current time step that doesn't line up with thermo? Also, is there a (synchronization) penalty  of `modify->addstep_compute()` vs doing `add_step` for one specific compute?

**Comment by sjplimp** [comment]
Re: penalty - I don't think so.  All the method does is perform an add_step for computes that were flagged as invoked since the clearstep() call.  The purpose of the pairing of clearstep() and addstep_compute() to bracket lots of chunks of code in LAMMPS is to free the calling code (or user) from needing to worry about which computes the bracketed code invoked.  E.g. by a fix or a dump or a variable or another compute.  So that the calling code can then set the next future step when energy/virial needs to be tallied for the computes that need it.

Re: another Python driver doing a run X, then asking for info on an arbitrary current timestep.  This is also a problem in an input script, say using a print command with a variable that accesses a compute between runs.   The "Variable Accuracy" section at the bottom of the variable doc page discusses possible work-arounds.

That said, I am not against adding a lib interface function or two, to let Python code directly set future timesteps for a compute to tally energy or virial quantities.

Two things to note.  Setting the compute tally flag near the beginning of a timestep is too late.  The Verlet::run() loop calls ev_set() at the very start of each timestep to set the tally flags passed to the Pair::compute().  Also, arbitrary Python code could run into the same issue the current C++ approach tries to solve.  Which is not easily knowing what compute(s) need their tally step 
set, since computes can be invoked indirectly via various operations which Python code could trigger in a manner similar to how an input script triggers it.

**Comment by rbberger** [comment]
Ok, I'll try to implement this.

**Comment by akohlmey** [comment]
@rbberger is this still missing something? or did PR #4380 implement everything needed and we can close this issue?
